File: mahjong_utils/models/furo.py
from abc import ABC
from typing import List, Sequence, Union

from pydantic.dataclasses import dataclass

from .mentsu import Shuntsu, Kotsu, Mentsu
from .tile import Tile, parse_tiles
from .tile_type import TileType


class Furo(Mentsu, ABC):
    """
    副露
    """

    pass


@dataclass(frozen=True)
class Chi(Shuntsu, Furo):
    """
    吃
    """
    pass


@dataclass(frozen=True)
class Pon(Kotsu, Furo):
    """
    碰
    """
    pass


@dataclass(frozen=True)
class Kan(Kotsu, Furo):
    """
    杠
    """

    ankan: bool

    def __repr__(self):
        if self.ankan:
            return f"0{self.tile.num}{self.tile.num}0{self.tile.tile_type}"
        else:
            return f"{self.tile.num}{self.tile.num}{self.tile.num}{self.tile.num}{self.tile.tile_type}"

    @property
    def tiles(self) -> List[Tile]:
        return [self.tile] * 4


def parse_furo(t: Union[Sequence[Tile], str]) -> Furo:
    if isinstance(t, str):
        t = parse_tiles(t)

    if len(t) == 3:
        if t[0] == t[1] == t[2]:
            return Pon(t[0])
        else:
            if t[0].tile_type == TileType.Z or t[1].tile_type == TileType.Z or t[2].tile_type == TileType.Z:
                raise ValueError(f"invalid tiles: {t}")
            t = sorted(t)
            if t[1] - t[0] == 1 and t[2] - t[1] == 1:
                return Chi(t[0])
            else:
                raise ValueError(f"invalid tiles: {t}")
    elif len(t) == 4:
        if t[0] == t[1] == t[2] == t[3]:
            return Kan(t[0], False)
        elif t[0] == t[3] and t[1] == t[2] and t[0].tile_type == t[1].tile_type and t[0].num == 0:
            #  0880p 这样的文本表示暗杠
            return Kan(t[1], True)

    raise ValueError(f"invalid tiles: {t}")

File: src/commonMain/kotlin/mahjongutils/models/Furo.kt
@file:OptIn(ExperimentalSerializationApi::class)

package mahjongutils.models

import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * 副露
 */
@Serializable
sealed interface Furo {
    /**
     * 获取副露的面子
     */
    fun asMentsu(): Mentsu

    companion object {
        /**
         * 根据给定牌构造副露
         * @param tiles 牌
         * @param ankan 是否为暗杠
         * @return 副露
         */
        operator fun invoke(tiles: List<Tile>, ankan: Boolean = false): Furo {
            if (tiles.size == 3) {
                if (tiles[0] == tiles[1] && tiles[1] == tiles[2]) {
                    return Pon(tiles[0])
                } else {
                    if (tiles.any { it.type == TileType.Z }) {
                        throw IllegalArgumentException("invalid tiles: ${tiles.toTilesString()}")
                    }

                    val tiles = tiles.sorted()
                    if (tiles[1].distance(tiles[0]) == 1 && tiles[2].distance(tiles[1]) == 1) {
                        return Chi(tiles[0])
                    } else {
                        throw IllegalArgumentException("invalid tiles: ${tiles.toTilesString()}")
                    }
                }
            } else if (tiles.size == 4) {
                if (tiles[0] == tiles[1] && tiles[1] == tiles[2] && tiles[2] == tiles[3]) {
                    return Kan(tiles[0], ankan)
                }
            }

            throw IllegalArgumentException("invalid tiles: ${tiles.toTilesString()}")
        }

        /**
         * 根据给定牌构造副露
         * @param text 牌的文本
         * @param ankan 是否为暗杠
         * @return 副露
         */
        operator fun invoke(text: String, ankan: Boolean = false): Furo {
            var ankan_ = ankan

            val tiles = if (text.length == 5 && text[0] == text[3] && text[0] == '0' && text[1] == text[2]) {
                ankan_ = true
                Tile.parseTiles("${text[1]}${text[1]}${text[1]}${text[1]}${text[4]}")
            } else {
                Tile.parseTiles(text)
            }

            return invoke(tiles, ankan_)
        }
    }
}

/**
 * 吃
 */
@Serializable
@SerialName("Chi")
data class Chi(
    /**
     * 吃成的顺子的第一张牌（如789m，tile应为7m）
     */
    val tile: Tile
) : Furo {
    override fun asMentsu(): Shuntsu {
        return Shuntsu(tile)
    }
}

/**
 * 碰
 */
@Serializable
@SerialName("Pon")
data class Pon(
    /**
     * 碰成的刻子的牌（如777s，tile应为7s）
     */
    val tile: Tile
) : Furo {
    override fun asMentsu(): Kotsu {
        return Kotsu(tile)
    }
}

/**
 * 杠
 */
@Serializable
@SerialName("Kan")
data class Kan(
    /**
     * 杠成的刻子的牌（如7777s，tile应为7s）
     */
    val tile: Tile,
    /**
     * 是否为暗杠
     */
    @EncodeDefault val ankan: Boolean = false
) : Furo {
    override fun asMentsu(): Kotsu {
        return Kotsu(tile)
    }
}
